File: libsaxsimage/saxsimage.f90
!
! Fortran Binding for libsaxsimage
! Copyright (C) 2011 Daniel Franke <dfranke@users.sourceforge.net>
!
! This file is part of libsaxsimage.
!
! libsaxsimage is free software: you can redistribute it 
! and/or modify it under the terms of the GNU Lesser General
! Public License as published by the Free Software Foundation,
! either version 3 of the License, or (at your option) any
! later version.
!
! libsaxsimage is distributed in the hope that it will be
! useful, but WITHOUT ANY WARRANTY; without even the implied
! warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
! PURPOSE. See the GNU Lesser General Public License for
! more details.
!
! You should have received a copy of the GNU Lesser General
! Public License along with libsaxsimage. If not,
! see <http://www.gnu.org/licenses/>.
!
MODULE saxsimage
  USE ISO_C_BINDING
  INTEGER, PARAMETER :: DBL = SELECTED_REAL_KIND(p=13, r=200)

  TYPE :: saxs_image
    TYPE(C_PTR) :: c_ptr
  END TYPE

  PRIVATE
  PUBLIC saxs_image
  PUBLIC saxs_image_create
  PUBLIC saxs_image_read
  PUBLIC saxs_image_free
  PUBLIC saxs_image_width
  PUBLIC saxs_image_height
  PUBLIC saxs_image_data

CONTAINS
  SUBROUTINE saxs_image_create(img)
    TYPE(saxs_image), INTENT(inout) :: img

    INTERFACE
      FUNCTION c_saxs_image_create() &
               BIND(C, NAME="saxs_image_create")
        IMPORT C_PTR
        TYPE(C_PTR) :: c_saxs_image_create
      END FUNCTION
    END INTERFACE

    img%c_ptr = c_saxs_image_create()
  END SUBROUTINE

  SUBROUTINE saxs_image_read(img, filename, fmt, status)
    TYPE(saxs_image), INTENT(inout)        :: img
    CHARACTER(len=*), INTENT(in)           :: filename
    CHARACTER(len=*), INTENT(in), OPTIONAL :: fmt
    INTEGER, INTENT(out), OPTIONAL         :: status

    INTERFACE
      FUNCTION c_saxs_image_read(img, filename, fmt) &
               BIND(C, NAME="saxs_image_read")
        IMPORT C_PTR, C_INT, C_CHAR
        INTEGER(C_INT)                :: c_saxs_image_read
        TYPE(C_PTR), VALUE            :: img
        CHARACTER(len=1, kind=C_CHAR) :: filename
        CHARACTER(len=1, kind=C_CHAR) :: fmt
      END FUNCTION
    END INTERFACE

    INTEGER :: result

    IF (PRESENT(fmt)) THEN
      result = c_saxs_image_read(img%c_ptr, TRIM(filename) // C_NULL_CHAR, TRIM(fmt) // C_NULL_CHAR)
    ELSE
      result = c_saxs_image_read(img%c_ptr, TRIM(filename) // C_NULL_CHAR, C_NULL_CHAR)
    END IF

    IF (PRESENT(status)) status = result
  END SUBROUTINE

  SUBROUTINE saxs_image_free(img)
    TYPE(saxs_image), INTENT(inout) :: img

    INTERFACE
      SUBROUTINE c_saxs_image_free(img) &
                 BIND(C, NAME="saxs_image_free")
        IMPORT C_PTR
        TYPE(C_PTR), VALUE :: img
      END SUBROUTINE
    END INTERFACE

    CALL c_saxs_image_free(img%c_ptr)
  END SUBROUTINE

  INTEGER FUNCTION saxs_image_width(img)
    TYPE(saxs_image), INTENT(inout) :: img

    INTERFACE
      FUNCTION c_saxs_image_width(img) &
                 BIND(C, NAME="saxs_image_width")
        IMPORT C_LONG, C_PTR
        INTEGER(C_LONG)    :: c_saxs_image_width
        TYPE(C_PTR), VALUE :: img
      END FUNCTION
    END INTERFACE

    saxs_image_width = c_saxs_image_width(img%c_ptr)
  END FUNCTION

  INTEGER FUNCTION saxs_image_height(img)
    TYPE(saxs_image), INTENT(inout) :: img

    INTERFACE
      FUNCTION c_saxs_image_height(img) &
                 BIND(C, NAME="saxs_image_height")
        IMPORT C_LONG, C_PTR
        INTEGER(C_LONG)    :: c_saxs_image_height
        TYPE(C_PTR), VALUE :: img
      END FUNCTION
    END INTERFACE

    saxs_image_height = c_saxs_image_height(img%c_ptr)
  END FUNCTION

  SUBROUTINE saxs_image_data(img, data)
    TYPE(saxs_image), INTENT(inout)                   :: img
    INTEGER, DIMENSION(:,:), ALLOCATABLE, INTENT(out) :: data

    INTEGER :: i, j, width, height

    INTERFACE
      FUNCTION c_saxs_image_value(img, x, y) &
                 BIND(C, NAME="saxs_image_value")
        IMPORT C_INT, C_LONG, C_PTR
        INTEGER(C_LONG)       :: c_saxs_image_value
        TYPE(C_PTR), VALUE    :: img
        INTEGER(C_INT), VALUE :: x, y
      END FUNCTION
    END INTERFACE

    width = saxs_image_width(img)
    height = saxs_image_width(img)

    ALLOCATE(data(width, height))

    DO i = 1, width
      DO j = 1, height
        data(i,j) = c_saxs_image_value(img%c_ptr, i, j)
      END DO
    ENDDO
  END SUBROUTINE
END MODULE

File: libsaxsimage/src/saxsimage.c
/*
 * Main API for SAXS image creation and access.
 * Copyright (C) 2009-2015 Daniel Franke <dfranke@users.sourceforge.net>
 *
 * This file is part of libsaxsdocument.
 *
 * libsaxsdocument is free software: you can redistribute it 
 * and/or modify it under the terms of the GNU Lesser General
 * Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any
 * later version.
 *
 * libsaxsdocument is distributed in the hope that it will be
 * useful, but WITHOUT ANY WARRANTY; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE. See the GNU Lesser General Public License for
 * more details.
 *
 * You should have received a copy of the GNU Lesser General
 * Public License along with libsaxsdocument. If not,
 * see <http://www.gnu.org/licenses/>.
 */

#include "saxsimage.h"
#include "saxsimage_format.h"
#include "saxsproperty.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <float.h>
#include <errno.h>
#include <assert.h>

struct saxs_image {
  char *image_filename;

  size_t image_width;
  size_t image_height;
  double *image_data;

  size_t image_frame_count;
  size_t image_current_frame;

  /* Callback functions to deal with the current format. */
  const saxs_image_format *image_format;

  saxs_property_list *image_properties;

  /* pre computed and cached values */
  int cache_valid;
  double cache_min_value, cache_max_value;
};


static void
saxs_image_update_cache(saxs_image *image) {
  if (image && !image->cache_valid) {
    size_t i, j;
    const size_t width = saxs_image_width(image);
    const size_t height = saxs_image_height(image);

    image->cache_min_value = DBL_MAX;
    image->cache_max_value = DBL_MIN;

    for (i = 0; i < width; ++i)
      for (j = 0; j < height; ++j) {
        double value = saxs_image_value(image, i, j);
        if (image->cache_min_value > value)
          image->cache_min_value = value;
        if (image->cache_max_value < value)
          image->cache_max_value = value;
      }

    image->cache_valid = 1;
  }
}



saxs_image*
saxs_image_create() {
  saxs_image *image = malloc(sizeof(saxs_image));
  if (!image)
    return NULL;

  image->image_filename      = NULL;
  image->image_width         = 0;
  image->image_height        = 0;
  image->image_data          = NULL;
  image->image_frame_count   = 0;
  image->image_current_frame = 0;
  image->cache_valid         = 0;
  image->cache_min_value     = DBL_MAX;
  image->cache_max_value     = DBL_MIN;
  image->image_format        = NULL;
  image->image_properties    = saxs_property_list_create();
  if (!image->image_properties) {
    free(image);
    return NULL;
  }

  return image;
}

saxs_image*
saxs_image_copy(saxs_image *image) {
  if (!image)
    return NULL;

  saxs_image *copy = malloc(sizeof(saxs_image));
  if (!copy)
    return NULL;

  if (image->image_filename) {
    copy->image_filename      = strdup(image->image_filename);
    if (!copy->image_filename) {
      saxs_image_free(copy);
      return NULL;
    }
  }
  copy->image_data          = NULL;
  copy->image_format        = NULL;
  copy->image_frame_count   = image->image_frame_count;
  copy->image_current_frame = image->image_current_frame;
  copy->cache_valid         = image->cache_valid;
  copy->cache_min_value     = image->cache_min_value;
  copy->cache_max_value     = image->cache_max_value;

  copy->image_properties    = saxs_property_list_create();
  if (!copy->image_properties) {
    saxs_image_free(copy);
    return NULL;
  }
  const saxs_property *prop;
  for (prop = saxs_property_list_first(image->image_properties);
       prop != NULL;
       prop = saxs_property_next(prop)) {
    saxs_property *prop_copy = saxs_property_create(saxs_property_name(prop), saxs_property_value(prop));
    if (!prop_copy) {
      saxs_image_free(copy);
      return NULL;
    }
    saxs_property_list_insert(copy->image_properties, prop_copy);
  }

  if (image->image_data) {
    saxs_image_set_size(copy, image->image_width, image->image_height,
                        image->image_frame_count, image->image_current_frame);
    if (!copy->image_data) {
      saxs_image_free(copy);
      return NULL;
    }

    memcpy(copy->image_data, image->image_data,
           image->image_width * image->image_height * sizeof(double));
  }

  assert(copy);
  if(image->image_filename){
    assert(copy->image_filename);
    assert(!strcmp(image->image_filename, copy->image_filename));
  }
  assert(copy->image_properties);
  assert(saxs_property_list_count(copy->image_properties) == saxs_property_list_count(image->image_properties));
  if (image->image_data)
    assert(copy->image_data);
  return copy;
}

int
saxs_image_read(saxs_image *image, const char *filename, const char *format) {
  assert(image);
  assert(filename);

  const saxs_image_format* handler = saxs_image_format_find(filename, format);
  if (!handler)
    return -1;
  if (!handler->read)
    return -2;

  free(image->image_filename);
  image->image_filename = strdup(filename);
  if (!image->image_filename)
    return -3;
  image->image_format   = handler;

  int res = image->image_format->read(image, filename, 1);

  if (res == 0) {
    assert(image->image_data);
    assert(image->image_frame_count > 0);
    assert(image->image_current_frame == 1);
    assert(image->image_height > 0);
    assert(image->image_width > 0);
  }
  return res;
}

int
saxs_image_write(saxs_image *image, const char *filename, const char *format) {
  assert(image);
  assert(filename);
  assert(image->image_data);
  assert(image->image_frame_count > 0);
  assert(image->image_height > 0);
  assert(image->image_width > 0);

  const saxs_image_format* handler = saxs_image_format_find(filename, format);
  if (!handler)
    return -1;
  if (!handler->write)
    return -2;

  free(image->image_filename);
  image->image_filename = strdup(filename);
  if (!image->image_filename)
    return -3;
  image->image_format   = handler;

  return image->image_format->write(image, filename);
}

void
saxs_image_free(saxs_image *image) {
  if (image) {
    if (image->image_filename)
      free(image->image_filename);

    if (image->image_data)
      free(image->image_data);

    saxs_property_list_free(image->image_properties);

    free(image);
  }
}

const char *
saxs_image_filename(saxs_image *image) {
  return image ? image->image_filename : NULL;
}

size_t
saxs_image_width(saxs_image *image) {
  return image ? image->image_width : 0;
}

size_t
saxs_image_height(saxs_image *image) {
  return image ? image->image_height : 0;
}

int
saxs_image_frame_count(saxs_image *image) {
  return image ? image->image_frame_count : 0;
}

int
saxs_image_current_frame(saxs_image *image) {
  return image ? image->image_current_frame : 0;
}

void
saxs_image_set_size(saxs_image *image, size_t width, size_t height,
                    size_t frame_count, size_t current_frame) {

  if (image->image_data)
    free(image->image_data);

  image->image_width         = width;
  image->image_height        = height;
  image->image_frame_count   = frame_count;
  image->image_current_frame = current_frame;
  image->image_data          = (double*) calloc(width * height, sizeof(double));
}

int
saxs_image_read_frame(saxs_image *image, size_t frame) {
  assert(image);
  assert(frame >= 1);

  if (frame > saxs_image_frame_count(image))
    return EINVAL;

  if (frame == image->image_current_frame)
    return 0;

  int res = image->image_format->read(image, image->image_filename, frame);

  assert(image->image_data);
  assert(frame == image->image_current_frame);
  return res;
}


double
saxs_image_value(saxs_image *image, int x, int y) {
  assert(image != NULL);
  assert(image->image_data != NULL);
  assert(x >= 0);
  assert(x < (signed)image->image_width);
  assert(y >= 0);
  assert(y < (signed)image->image_height);

  return image->image_data[y * image->image_width + x];
}

void
saxs_image_set_value(saxs_image *image, int x, int y, double value) {
  assert(image != NULL);
  assert(image->image_data != NULL);
  assert(x >= 0);
  assert(x < (signed)image->image_width);
  assert(y >= 0);
  assert(y < (signed)image->image_height);

  image->image_data[y * image->image_width + x] = value;

  if (image->cache_valid)
    image->cache_valid = 0;
}


double
saxs_image_value_min(saxs_image *image) {
  if (image) {
    saxs_image_update_cache(image);
    return image->cache_min_value;

  } else
    return 0.0;
}

double
saxs_image_value_max(saxs_image *image) {
  if (image) {
    saxs_image_update_cache(image);
    return image->cache_max_value;

  } else
    return 0.0;
}



saxs_property*
saxs_image_add_property(saxs_image *image, const char *name, const char *value) {
  if (!image)
    return NULL;

  saxs_property *property = saxs_property_create(name, value);
  if (!property)
    return NULL;

  saxs_property_list_insert(image->image_properties, property);
  return property;
}

int
saxs_image_property_count(saxs_image *image) {
  return saxs_property_list_count(image->image_properties);
}

saxs_property*
saxs_image_property_first(saxs_image *image) {
  return saxs_property_list_first(image->image_properties);
}

saxs_property*
saxs_image_property_find_first(saxs_image *image, const char *name) {
  return saxs_property_list_find_first(image->image_properties, name);
}
